src/lane_keep_assist/scripts/lane_detect.py

```diff
@@ -0,0 +1,257 @@
+    line_segments = cv.HoughLinesP(edges, rho, angle, min_threshold,
+                                   np.array([]), minLineLength=HOUGH_MIN_LINE_LENGTH, maxLineGap=HOUGH_MAX_LINE_GAP)
+
+    return line_segments
+
+
+# Detect the hough lines, different methods to try using different parameters for segmented image
+def hough_line_detection_segmentation(edges: np.ndarray):
+    rho = 1  # distance precision in pixel, i.e. 1 pixel
+    angle = np.pi / 180  # angular precision in radian, i.e. 1 degree
+    min_threshold = HOUGH_MIN_VOTES_SEGMENTED  # minimal of votes
+    line_segments = cv.HoughLinesP(edges, rho, angle, min_threshold,
+                                   np.array([]), minLineLength=HOUGH_MIN_LINE_LENGTH_SEGMENTED, maxLineGap=HOUGH_MAX_LINE_GAP_SEGMENTED)
+
+    return line_segments
+
+
+# Get two points from a line
+def make_points(frame: np.ndarray, line: np.ndarray) -> List[List[float]]:
+    height, width, _ = frame.shape
+    slope, intercept = line
+    y1 = height  # bottom of the frame
+    y2 = int(y1 * 1 / 2)  # make points from middle of the frame down
+
+    # bound the coordinates within the frame
+    try:
+        x1 = int((y1 - intercept) / slope)
+        if not 0 <= x1 <= width:
+            if slope < 0:
+                x1 = 0
+            else:
+                x1 = width
+            y1 = int(x1*slope + intercept)
+        x2 = int((y2 - intercept) / slope)
+        if not 0 <= x2 <= width:
+            if slope > 0:
+                x2 = 0
+            else:
+                x2 = width
+            y2 = int(x2*slope + intercept)
+        return [x1, y1, x2, y2]
+    except OverflowError as e:
+        return [0, 0, 0, 0]
+
+
+# Divide the lanes into left and right segments
+def average_slope_intercept(frame: np.ndarray, line_segments: np.ndarray):
+    lane_lines = []
+    if line_segments is None:
+        return lane_lines
+
+    height, width, _ = frame.shape
+    left_fit = []
+    right_fit = []
+
+    boundary = width * 1 / 2
+
+    for line_segment in line_segments:
+        for x1, y1, x2, y2 in line_segment:
+            if x1 == x2:
+                continue
+            fit = np.polyfit((x1, x2), (y1, y2), 1)
+            slope = fit[0]
+            intercept = fit[1]
+
+            # Try and remove more horizontal hough lines that are likely not lanes
+            if 0.1 > slope > -0.1:
+                continue
+
+            if x1 < boundary and x2 < boundary:
+                left_fit.append((slope, intercept))
+            elif x1 > boundary and x2 > boundary:
+                right_fit.append((slope, intercept))
+            else:
+                continue
+
+    left_fit_average = np.average(left_fit, axis=0)
+    if len(left_fit) > 0:
+        lane_lines.append(make_points(frame, left_fit_average))
+
+    right_fit_average = np.average(right_fit, axis=0)
+    if len(right_fit) > 0:
+        lane_lines.append(make_points(frame, right_fit_average))
+
+    return lane_lines
```

Comment:
If it returns a single lane line, does the caller know if its a right or left one?

Reply:
yeah, that will be indicated by the status, I'll make an enum for that to make it more clear.